source/common/upstream/subset_lb.cc

```diff
@@ -215,6 +215,21 @@ void SubsetLoadBalancer::processSubsets(
 // new subsets as necessary.
 void SubsetLoadBalancer::update(uint32_t priority, const HostVector& hosts_added,
                                 const HostVector& hosts_removed) {
+  // It's possible that metadata changed, without hosts being added nor removed.
+  // If so we need to add any new subsets, remove unused ones, and regroup hosts into
+  // the right subsets.
+  if (!hosts_added.size() && !hosts_removed.size()) {
```

Comment:
Is it possible that this logic doesn't work if the endpoints being updated had some additions/removals with other endpoints remaining the same but with only metadata changes?

Reply:
Not sure I follow what combination of changes you are referring to.. Are you saying:

* metadata changes for some of the existing endpoints
* new endpoints are added (they must be different from the existing one, given how we operate now)
* some endpoints are removed 

?

Reply:
Easiest way to verify is another test-case, but I believe the question here is:
If there were additions/removals for some endpoints, and metadata-only changes for others 
  1) this block would not be taken, since the size _did_ change
  2) the hosts added/removed would be taken care of down below
  3) the metadata-only changes are not addressed

Reply:
Ah, now I got what you were asking thanks to @derekargueta. If there's an add/removal _and_ metadata changes for existing endpoints we won't miss it. Because as mentioned above, we refilter hosts in `SubsetLoadBalancer::HostSubsetImpl::update()`...

Let me throw in another unit test and add a comment about this.